File: lib/src/main/java/com/darklycoder/download/info/TaskStatusInfo.java
package com.darklycoder.download.info;

import com.darklycoder.download.interfaces.IStatusListener;

/**
 * 下载任务状态信息
 *
 * <pre>
 *     1、回调单个任务下载状态：进度、下载完成(失败/成功)
 *     2、回调总体下载进度
 * </pre>
 */
public class TaskStatusInfo {

    // 总数
    private long count;
    // 成功个数
    private long successCount;
    // 失败个数
    private long failCount;
    private IStatusListener listener;

    public TaskStatusInfo(long count, IStatusListener listener) {
        this.count = count;
        this.listener = listener;
    }

    public void onCellProgress(TaskCellInfo info, float progress) {
        if (null == listener) {
            return;
        }

        listener.onCellProgress(info, progress);
    }

    public void onCellSuccess(TaskCellInfo info) {
        successCount++;
        if (null == listener) {
            return;
        }

        listener.onCellSuccess(info);
        onTotalProgress();
    }

    public void onCellFail(TaskCellInfo info) {
        failCount++;
        if (null == listener) {
            return;
        }

        listener.onCellFail(info);
        onTotalProgress();
    }

    /**
     * 是否全部执行完毕
     */
    public boolean isComplete() {
        return count == successCount + failCount;
    }

    /**
     * 获取整体下载进度
     */
    private float getProgress() {
        return (successCount + failCount) * 1F / count;
    }

    private void onTotalProgress() {
        listener.onTotalProgress(getProgress(), count, successCount, failCount);
    }

}

File: lib/src/main/java/com/darklycoder/download/info/TaskStatusInfo.kt
package com.darklycoder.download.info

import com.darklycoder.download.interfaces.IStatusListener

/**
 * 下载任务状态信息
 *
 * <pre>
 * 1、回调单个任务下载状态：进度、下载完成(失败/成功)
 * 2、回调总体下载进度
 * </pre>
 */
class TaskStatusInfo(
        // 总数
        private val count: Int,
        // 回调
        private val listener: IStatusListener? = null
) {

    // 成功个数
    private var successCount: Int = 0
    // 失败个数
    private var failCount: Int = 0

    /**
     * 是否全部执行完毕
     */
    val isComplete: Boolean
        get() = count == successCount + failCount

    /**
     * 获取整体下载进度
     */
    private val progress: Float
        get() = (successCount + failCount) * 1f / count

    fun onCellProgress(info: TaskCellInfo, progress: Float) {
        listener?.onCellProgress(info, progress)
    }

    fun onCellSuccess(info: TaskCellInfo) {
        successCount++

        listener?.onCellSuccess(info)
        onTotalProgress()
    }

    fun onCellFail(info: TaskCellInfo) {
        failCount++

        listener?.onCellFail(info)
        onTotalProgress()
    }

    private fun onTotalProgress() {
        listener?.onTotalProgress(progress, count, successCount, failCount)
    }

}
